### ISSUE TITLE
Timeout when Weblate updates a very large component

### ISSUE THREAD

**Issue by endervad** [WeblateOrg/weblate#7382]
### Describe the issue

I have a component which contains more than 42,000 strings. And there are 6 languages, so it results in more than 250,000 strings overall. When trying to create such component or update it by changing a setting, Weblate timeouts after an hour and in the end doesn't do anything. Reducing amount of languages to 4 helps, but ultimately it doesn't solve the problem. It happens both from the web, and after the REST request.

### I already tried

- [X] I've read and searched [the documentation](https://docs.weblate.org/).
- [X] I've searched for similar issues in this repository.

### Steps to reproduce the behavior

1. Add the attached component from ZIP file (I'm using remote repo but it doesn't matter)
[ffxiv-translation-enpcresident.zip](https://github.com/WeblateOrg/weblate/files/8246183/ffxiv-translation-enpcresident.zip)
2. Weblate will work for an hour and then throw a 504 error page

### Expected behavior

* Faster (preferably much faster) adding/updating component. I noticed that during this process only one thread is utilized. I'm pretty sure that with multithreading this wouldn't be an issue
* Having an option to remove timeout altogether so that it would run for as long as it needs to. Even though it's a workaround rather than solution, it'd be nice to have such option until solution is found

### How do you run Weblate?

Docker container

### Weblate versions

 * Weblate: 4.11.2
 * Django: 4.0.3
 * siphashc: 2.1
 * translate-toolkit: 3.6.0
 * lxml: 4.7.1
 * Pillow: 9.0.1
 * bleach: 4.1.0
 * python-dateutil: 2.8.2
 * social-auth-core: 4.2.0
 * social-auth-app-django: 5.0.0
 * django-crispy-forms: 1.14.0
 * oauthlib: 3.2.0
 * django-compressor: 3.1
 * djangorestframework: 3.13.1
 * django-filter: 21.1
 * django-appconf: 1.0.5
 * user-agents: 2.2.0
 * filelock: 3.6.0
 * jellyfish: 0.9.0
 * openpyxl: 3.0.9
 * celery: 5.2.3
 * kombu: 5.2.3
 * translation-finder: 2.12
 * weblate-language-data: 2022.2
 * html2text: 2020.1.16
 * pycairo: 1.20.1
 * pygobject: 3.42.0
 * diff-match-patch: 20200713
 * requests: 2.27.1
 * django-redis: 5.2.0
 * hiredis: 2.0.0
 * sentry_sdk: 1.5.6
 * Cython: 0.29.28
 * misaka: 2.1.1
 * GitPython: 3.1.27
 * borgbackup: 1.2.0
 * pyparsing: 3.0.7
 * pyahocorasick: 1.4.4
 * python-redis-lock: 3.7.0
 * charset-normalizer: 2.0.12
 * Python: 3.10.2
 * Git: 2.30.2
 * psycopg2: 2.9.3
 * psycopg2-binary: 2.9.3
 * phply: 1.2.5
 * ruamel.yaml: 0.17.21
 * tesserocr: 2.5.2
 * akismet: 1.1
 * boto3: 1.21.12
 * zeep: 4.1.0
 * aeidon: 1.10.1
 * iniparse: 0.5
 * mysqlclient: 2.1.0
 * Mercurial: 6.1
 * git-svn: 2.30.2
 * git-review: 2.2.0
 * Redis server: 6.2.6
 * PostgreSQL server: 13.6
 * Database backends: django.db.backends.postgresql
 * Cache backends: default:RedisCache, avatar:FileBasedCache
 * Email setup: django.core.mail.backends.smtp.EmailBackend: smtp.gmail.com
 * OS encoding: filesystem=utf-8, default=utf-8
 * Celery: redis://cache:6379/1, redis://cache:6379/1, regular
 * Platform: Linux 5.10.60.1-microsoft-standard-WSL2 (x86_64)

### Weblate deploy checks

```shell
System check identified some issues:

INFOS:
?: (weblate.I021) Error collection is not set up, it is highly recommended for production use
        HINT: https://docs.weblate.org/en/weblate-4.11.2/admin/install.html#collecting-errors
?: (weblate.I028) Backups are not configured, it is highly recommended for production use
        HINT: https://docs.weblate.org/en/weblate-4.11.2/admin/backup.html

System check identified 2 issues (1 silenced).
```

### Additional context

CPU: Intel Core i9-9900K (8 @ 3.6GHz with HT & up to 5GHz)
RAM: Dynamic for Weblate container, it eats up to 25Gb as necessary (`docker stats` shows that it's using much less but that's not on Weblate)

**Comment by nijel** [comment]
Creation happens in the background when using UI, so it won't timeout. API is currently synchronous in this. Changes should be way faster as only changed strings will be processed.

**Comment by endervad** [comment]
I tried uploading a file with updated strings into the attached big component and watching logs. I've found out that Postgres generates a massive amount of the same error:
```
2022-03-03 19:01:27.796 UTC [27949] ERROR:  duplicate key value violates unique constraint "trans_unit_translation_id_id_hash_ab945589_uniq"
2022-03-03 19:01:27.796 UTC [27949] DETAIL:  Key (translation_id, id_hash)=(22353, -2040007773529392207) already exists.
2022-03-03 19:01:27.796 UTC [27949] STATEMENT:  INSERT INTO "trans_unit" ("translation_id", "id_hash", "location", "context", "note", "flags", "source", "previous_source", "target", "state", "original_state", "details", "position", "num_words", "priority", "pending", "timestamp", "extra_flags", "explanation", "variant_id", "source_unit_id") VALUES (22353,  -2040007773529392207, '', '', '', '', 'Rockbreaker', '', '', 0, 0, '{"add_unit": true}', 20, 1, 100, true, '2022-03-03T19:01:27.796408+00:00'::timestamptz, '', '', NULL, 7115345) RETURNING "trans_unit"."id"
2022-03-03 19:01:27.801 UTC [27949] ERROR:  duplicate key value violates unique constraint "trans_unit_translation_id_id_hash_ab945589_uniq"
2022-03-03 19:01:27.801 UTC [27949] DETAIL:  Key (translation_id, id_hash)=(22350, -7008526211942312052) already exists.
2022-03-03 19:01:27.801 UTC [27949] STATEMENT:  INSERT INTO "trans_unit" ("translation_id", "id_hash", "location", "context", "note", "flags", "source", "previous_source", "target", "state", "original_state", "details", "position", "num_words", "priority", "pending", "timestamp", "extra_flags", "explanation", "variant_id", "source_unit_id") VALUES (22350,  -7008526211942312052, '', '', '', '', 'Dragon Kick', '', '', 0, 0, '{"add_unit": true}', 20, 2, 100, true, '2022-03-03T19:01:27.801503+00:00'::timestamptz, '', '', NULL, 7115396) RETURNING "trans_unit"."id"
2022-03-03 19:01:27.804 UTC [27949] ERROR:  duplicate key value violates unique constraint "trans_unit_translation_id_id_hash_ab945589_uniq"
2022-03-03 19:01:27.804 UTC [27949] DETAIL:  Key (translation_id, id_hash)=(22354, -7008526211942312052) already exists.
2022-03-03 19:01:27.804 UTC [27949] STATEMENT:  INSERT INTO "trans_unit" ("translation_id", "id_hash", "location", "context", "note", "flags", "source", "previous_source", "target", "state", "original_state", "details", "position", "num_words", "priority", "pending", "timestamp", "extra_flags", "explanation", "variant_id", "source_unit_id") VALUES (22354,  -7008526211942312052, '', '', '', '', 'Dragon Kick', '', '', 0, 0, '{"add_unit": true}', 20, 2, 100, true, '2022-03-03T19:01:27.804270+00:00'::timestamptz, '', '', NULL, 7115396) RETURNING "trans_unit"."id"
2022-03-03 19:01:27.807 UTC [27949] ERROR:  duplicate key value violates unique constraint "trans_unit_translation_id_id_hash_ab945589_uniq"
2022-03-03 19:01:27.807 UTC [27949] DETAIL:  Key (translation_id, id_hash)=(22351, -7008526211942312052) already exists.
2022-03-03 19:01:27.807 UTC [27949] STATEMENT:  INSERT INTO "trans_unit" ("translation_id", "id_hash", "location", "context", "note", "flags", "source", "previous_source", "target", "state", "original_state", "details", "position", "num_words", "priority", "pending", "timestamp", "extra_flags", "explanation", "variant_id", "source_unit_id") VALUES (22351,  -7008526211942312052, '', '', '', '', 'Dragon Kick', '', '', 0, 0, '{"add_unit": true}', 20, 2, 100, true, '2022-03-03T19:01:27.806840+00:00'::timestamptz, '', '', NULL, 7115396) RETURNING "trans_unit"."id"
2022-03-03 19:01:27.809 UTC [27949] ERROR:  duplicate key value violates unique constraint "trans_unit_translation_id_id_hash_ab945589_uniq"
2022-03-03 19:01:27.809 UTC [27949] DETAIL:  Key (translation_id, id_hash)=(22352, -7008526211942312052) already exists.
2022-03-03 19:01:27.809 UTC [27949] STATEMENT:  INSERT INTO "trans_unit" ("translation_id", "id_hash", "location", "context", "note", "flags", "source", "previous_source", "target", "state", "original_state", "details", "position", "num_words", "priority", "pending", "timestamp", "extra_flags", "explanation", "variant_id", "source_unit_id") VALUES (22352,  -7008526211942312052, '', '', '', '', 'Dragon Kick', '', '', 0, 0, '{"add_unit": true}', 20, 2, 100, true, '2022-03-03T19:01:27.809241+00:00'::timestamptz, '', '', NULL, 7115396) RETURNING "trans_unit"."id"
```
These CSV files were imported before this fix - #7247, and I see that Postgres complains about key value. In my project basically all of the components have the duplicate keys, and they are just hex numbers: 0x00000000, 0x00000001, 0x00000002, 0x00000003, etc. It also looks like it refers to strings in a different component of the same linked repo ("Dragon Kick" doesn't exist in current component, but exists in another). However I have disabled translation propagation for all components.
Could this happen because I didn't reimport CSVs after the mentioned fix, or the issue lies somewhere else entirely?

**Comment by endervad** [comment]
> Creation happens in the background when using UI, so it won't timeout. API is currently synchronous in this. Changes should be way faster as only changed strings will be processed.

I understand about API.
About web though, I noticed that it starts working in background after the component "entry" was created, when it shows live log with progress for each language file. I'm guessing you refer to this when you said it happens in background. However in my case, I don't even get to this page in time with a huge component like the one I attached in the issue. After pressing "Save" on "Add new translation component" page, the browser keeps waiting for the answer from Weblate. And this is where it timeouts after an hour.
Also, adding to my comment above, Weblate timeouted with 504 again after an hour. None of the strings were changed in the process.

**Comment by endervad** [comment]
Following the advice of disabling glossaries, I tried updating the big component too. In this case I tried to switch off the option "Send on commit". In log file I pressed the "Save" button at 17:22, component is called `enpcresident`. After that it seems like nothing related to this task was logged until after an hour when I got 504 error.

[weblate-component-update.zip](https://github.com/WeblateOrg/weblate/files/8404853/weblate-component-update.zip)
